---
##### Name: Luke Voyer  
##### Major: IT
##### Creation Date: 12/2/2020
##### Due Date: 12/9/2020
##### Course: Advanced Scientific Programming
##### Professor Name: Dr. Schweisinger
##### Assignment: Final Project
##### Filename: main.ipnyb
---

# Introduction
This dataet was obtained April through May of 2015. This was the time where this app was expanding most rapidly. 
At that time the iOS version of the Lovoo app was in version 3. This data was recorded using the iOS version 3 of the Lovoo app

Below are two datasets imported from the [Lovoo](https://www.lovoo.com/) dating app including dating profiles from exlusively 
**female** users, including data such as the gender they're looking for, number of pictures on their profile, country, age, etc. (Dating App Users & Stats - Lovoo v3 - dataset by jfreex, 2020)

In [2]:
import pandas as pd

# Data points available in lovoo_v3_users_api-results.csv not available in
# lovoo_v3_users_instances.csv:
#      genderlooking, user pseudo, counts_fans, counts_g, country, country_1
#      distance, isinfluencer, isnew, shareprofileenabled, birthd, freetext
user_api_results = pd.read_csv('lovoo_v3_users_api-results.csv')

# Any other field that is not in user_instances is in user_api_results,
# and if a user_api_results field is not used, user_instances is used.
user_instances = pd.read_csv('lovoo_v3_users_instances.csv')

The above datasets describe slightly different data fields. 'lovoo_v3_users_api-results.csv' describes more than 'lovoo_v3_users_instances.csv' in that it has all the fields that 'lovoo_v3_users_instances.csv' has, plus the fields 'genderlooking', 'isnew', and 'birthd' that describe the gender a person is looking for in a relationship, whether a profile is considered 'new', or a user's birthday. 

All other fields are identical and any fields that aren't unique to 'lovoo_v3_users_api-results.csv' will be pulled from 'lovoo_v3_users_instances.csv' henceforth.

In [3]:
# Counter variable for those under 20
under_20 = 0

# For loop to count those under 20
for a in user_instances['age']:
    if (a < 20):
        under_20 += 1;

# Print debug results
print("Quantatative:")
print("Number of observed female Lovoo users who are under 20 years old: " + str(under_20))

Quantatative:
Number of observed female Lovoo users who are under 20 years old: 369


The above steps takes the age field from the 'user_instances' table and counts those who are under the age of 20.
Stepping through the for loop that accounts for each age of each female signed up for Lovoo, it runs the if statement that checks if that age is less than 20, and if they are...one is added to the count of those who **are** under 20

In [4]:
# Number of straight (F-M only), bisexual users (F-F or F-M), or females looking for friends (F-friend) on Lovoo:
bi_relat = 0
straight_relat = 0
not_looking = 0

# Sort and count those looking for bisexual and straight relationships
# as well as those not looking for relationships
for g in user_api_results['genderLooking']:
    if (g == 'both'):
        bi_relat += 1;
    elif (g == 'M'):
        straight_relat += 1;
    elif (g == 'none'):
        not_looking += 1;

# Print debug results
print("Categorical:")
print("Number of observed female Lovoo users who are looking for a straight relationship: " + str(straight_relat))
print("Number of observed female Lovoo users who are looking for a bisexual relationship: " + str(bi_relat))
print("Number of observed female Lovoo users who are not looking for a relationship: " + str(not_looking))

Categorical:
Number of observed female Lovoo users who are looking for a straight relationship: 3561
Number of observed female Lovoo users who are looking for a bisexual relationship: 21
Number of observed female Lovoo users who are not looking for a relationship: 366


The above steps takes the unique 'genderLooking' field from the 'user_api_results' table and counts those who are either looking to be in a straight or bisexual relationship, or not looking for one at all. Stepping through the for loop that catagorizes based on thier preference, it runs the if statement that checks if that which 'bin' to put them in... and one is added to that count.

In [4]:
# Variables for the mean of picture count and picture count grouped by ages of women 
sr_mean = user_instances.groupby(['age'])['counts_pictures'].mean()
sr_count = user_instances.groupby(['age'])['counts_pictures'].count()

print("Grouping:")
#Remove ages that are poorly accounted for: (Count of 1 and ony 1 instance)
print("Ages where number of females observed is poorly accounted for (Count of 1): ")
for i in sr_count[sr_count == 1].index:
    print(i, end=" ")
    sr_mean = sr_mean.drop(i)

print("\n")

# Debug print for average count of photos for females of each observed ages
counter  = 0;
for i in sr_mean.index:
    print("Average photos added to a profile for females age " + str(i), end=": ")
    print(int(sr_mean.values[counter]))
    counter += 1;

Grouping:
Ages where number of females observed is poorly accounted for (Count of 1): 
18 27 28 

Average photos added to a profile for females age 19: 5
Average photos added to a profile for females age 20: 5
Average photos added to a profile for females age 21: 6
Average photos added to a profile for females age 22: 4
Average photos added to a profile for females age 23: 4
Average photos added to a profile for females age 24: 4
Average photos added to a profile for females age 25: 4
Average photos added to a profile for females age 26: 3


The above steps takes the 'age' field from the 'user_instances' table and groups the count of photos in a profile per age. The first for loop purges data that is 'insufficiently recorded', or data that only has one data point. 

Stepping through the for loop that catagorizes based on thier preference, it runs the if statement that checks if that which 'bin' to put them in... and one is added to that count.

In [5]:
# Join on user_api_results and _fromInstances
api_join_instances = user_api_results.join(user_instances, lsuffix='_fromAPI', rsuffix='_fromInstances')

# Varaibles to store profile visits from those who do and don't have bio's on their profiles
count_bio_null = api_join_instances['counts_profileVisits_fromAPI'].loc[api_join_instances['whazzup_fromInstances'].isnull()].mean()
count_bio_not_null = api_join_instances['counts_profileVisits_fromAPI'].loc[api_join_instances['whazzup_fromInstances'].notnull()].mean()

print("Joining:")
# Debug prints to determine whether those who do have bio's or those who don't get more visits
if (count_bio_null > count_bio_not_null):
    print("Those who have bio's in their profiles have more views on their profiles than those who don't.");
else:
    print("Those who don't have bio's in their profiles have more views on their profiles than those who do.");

Joining:
Those who don't have bio's in their profiles have more views on their profiles than those who do.


The above steps takes the joined 'user_api_results' and 'user_instances' tables, then counts both profile visits those who have bio's and those who don't.

Then the two are compared are an output is shown basedm on whether or not those who have bio's or those who don't get more profile views.

# Conclusion:

The investigation shows multiple aspects of this dataset.

This investigation shows that there are 369 users of Lovoo under 20. There are 3561 observed female Lovoo users who are looking for a straight relationship, 21 looking for a bisexual relationship, and 366 who are just looking for friends. The average amount of photos per profile can range from 3 to 6 photos on average, with 21 year olds posting the most, to 26 year olds posting the least. Lastly, this investigation shows that surprisingly, that those who don't have bio's on their profiles actually get more views than those who do.

In conclusion, this dataset contains multiple types of data, with multiple different facets and values, but each find just happend to be as interesting as the last.

# Works Cited:

* Dating App Users &amp; Stats - Lovoo v3 - dataset by jfreex. (2020, September 02). Retrieved December 09, 2020, from https://data.world/jfreex/dating-app-user-profiles-stats-lovoo-v3